### PR TITLE
Index Period fields as full datetimes

### LINKED ISSUE #1995: Index `Period` search parameters as `datetimes`
Right now, when we parse a `date` search parameter type, we look at the StructureDefinition to see if it the search should be stored as a DATE or DATETIME column. We need to modify this condition to handle everything that is not `PropertyType.date` as  a DateTime

```
function getSearchParameterType(searchParam: SearchParameter, propertyType?: PropertyType): SearchParameterType {
  let type = SearchParameterType.TEXT;
  switch (searchParam.type) {
    case 'date':
      if (propertyType === PropertyType.dateTime || propertyType === PropertyType.instant) {
        type = SearchParameterType.DATETIME;
      } else {
        type = SearchParameterType.DATE;
      }
      break;
  // ...
 }
}
```

### PR DESCRIPTION
**Pull request by mattwiller**
Fixes #1995

### CODE REVIEW

**Comment by rahul1** [comment]
This PR will probably also require a DB migration to convert existing `date` search columns to `dateTimes`

**Comment by tonylee80** [comment]
You have a failed unit test
```
@medplum/server:test: FAIL src/fhir/repo.test.ts (7.794 s)
@medplum/server:test:   ● FHIR Repo › Encounter.period date search
@medplum/server:test: 
@medplum/server:test:     expect(received).toBeTruthy()
@medplum/server:test: 
@medplum/server:test:     Received: false
@medplum/server:test: 
@medplum/server:test:       2797 |     });
@medplum/server:test:       2798 |
@medplum/server:test:     > 2799 |     expect(bundleContains(bundle, e)).toBeTruthy();
@medplum/server:test:            |                                       ^
@medplum/server:test:       2800 |   });
@medplum/server:test:       2801 |
@medplum/server:test:       2802 |   test('Condition.code system search', async () => {
@medplum/server:test: 
@medplum/server:test:       at toBeTruthy (src/fhir/repo.test.ts:2799:39)
@medplum/server:test:       at call (src/fhir/repo.test.ts:2:1)
@medplum/server:test:       at Generator.tryCatch (src/fhir/repo.test.ts:2:1)
@medplum/server:test:       at Generator._invoke [as next] (src/fhir/repo.test.ts:2:1)
@medplum/server:test:       at asyncGeneratorStep (src/fhir/repo.test.ts:2:1)
@medplum/server:test:       at asyncGeneratorStep (src/fhir/repo.test.ts:2:1)
```

https://github.com/medplum/medplum/actions/runs/4920121936/jobs/8788577634?pr=1996#step:6:1586

File: packages/core/src/search/details.test.ts [hunk 1]
```diff
@@ -120,22 +120,6 @@ describe('SearchParameterDetails', () => {
     expect(() => getSearchParameterDetails('Patient', missingExpressionParam)).toThrow();
   });
 
-  test('Observation-value-date', () => {
```
Comment (codyebberson) [inline packages/core/src/search/details.test.ts:123]: Out of curiosity, why delete this test?

**Comment by codyebberson** [comment]
FYI: one of my concerns was `SearchParameter` resources that point to a `[x]` choice of type field that can be `date` or other.  I did a quick check, and happily that is not the case.

There are two near misses:
1. `Goal` `start-date` refers to `Goal.start`, which can be either `date` or `CodeableConcept`.  But the `SearchParameter.expresssion` explicitly requires `as date`: `(Goal.start as date)`
2. Likewise `Goal` `target-date` refers to `Goal.target.due`, which can be either `date` or `Duration`.  But the `SearchParameter.expresssion` explicitly requires `as date`: `(Goal.target.due as date)`

So I think everything checks out.

File: packages/server/src/fhir/repo.ts [hunk 2]
```diff
@@ -1787,6 +1785,9 @@ export class Repository extends BaseRepository implements FhirRepository {
       } catch (ex) {
         // Silent ignore
       }
+    } else if (typeof value === 'object' && 'start' in value) {
+      // Can be a Period
+      return this.buildDateColumn(value.start);
```
Comment (codyebberson) [inline packages/server/src/fhir/repo.ts:1790]: This needs to be `buildDateTimeColumn`

**Review by codyebberson — APPROVED** [review]

**Pull request merged**